module-bsp/board/rt1051/puretx/bsp/keyboard/keyboard.cpp

```diff
@@ -67,24 +68,47 @@ namespace bsp::keyboard
                 }
             }
 
-            gpio->EnableInterrupt(1U << static_cast<uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON));
+            gpio->EnableInterrupt(1U << static_cast<std::uint32_t>(BoardDefinitions::KEYBOARD_RF_BUTTON));
         }
+    } // anonymous namespace
 
-        void clearAllIRQs()
-        {
-            std::uint8_t clearIRQs = 0xFF;
-            i2cAddr.subAddress     = REG_INT_STAT;
-            i2c->Write(i2cAddr, reinterpret_cast<std::uint8_t *>(&clearIRQs), 1);
+    bool write(const drivers::I2CAddress &addr, const std::uint8_t *txBuff, const std::size_t size)
+    {
+        for (auto i = 0; i < maxTransmissionRetry; ++i) {
+            auto ret = i2c->Write(addr, txBuff, size);
+            if (ret == size) {
+                return true;
+            }
         }
+        LOG_ERROR("Can't write data to keyboard. I2C error: %ld", i2c->GetLastError());
+        return false;
```

Comment:
So at this point the keyboard will stop to respond, correct?
What should we do in this case?
Shouldn't we e.g. restart the phone if that happens since the phone is going to be unusable anyway?

Reply:
First of all, I want to register transmission errors on i2c bus. I haven't considered restarting the phone yet, but I am afraid if i2c bus is broken, the firmware can restart all the time.

Reply:
I'm not sure whether soft-restarting the phone is really a good solution here - when I encountered this bug, the only way to make phone responsive again was to remove battery, what physically "restarted" all ICs to initial state. I assume there was some unrecoverable condition on the bus that couldn't be fixed by OS restart, e.g. some IC was holding SDA low, what rendered further communication impossible.

Reply:
It is also hard to reproduce so at this point I prefer to gather data and then try to find the reason for this fault.